src/IO/S3/PocoHTTPClient.cpp

```diff
@@ -314,30 +291,81 @@ void PocoHTTPClient::addMetric(const Aws::Http::HttpRequest & request, S3MetricT
 }
 
-void PocoHTTPClient::addLatency(const Aws::Http::HttpRequest & request, S3LatencyType type, LatencyBuckets::Count amount) const
+void PocoHTTPClient::observeLatency(const Aws::Http::HttpRequest & request, S3LatencyType type, Histogram::Value latency) const
 {
-    if (amount == 0)
+    if (latency == 0)
         return;
 
-    static const LatencyBuckets::Event events_map[static_cast<size_t>(S3LatencyType::EnumSize)][static_cast<size_t>(S3MetricKind::EnumSize)] = {
-        {LatencyBuckets::S3FirstByteReadAttempt1Microseconds, LatencyBuckets::S3FirstByteWriteAttempt1Microseconds},
-        {LatencyBuckets::S3FirstByteReadAttempt2Microseconds, LatencyBuckets::S3FirstByteWriteAttempt2Microseconds},
-        {LatencyBuckets::S3FirstByteReadAttemptNMicroseconds, LatencyBuckets::S3FirstByteWriteAttemptNMicroseconds},
-        {LatencyBuckets::S3ConnectMicroseconds, LatencyBuckets::S3ConnectMicroseconds},
-    };
+    if (type == S3LatencyType::Connect)
+    {
+        const Histogram::Buckets connect_buckets = {100, 1000, 10000, 100000, 200000, 300000, 500000, 1000000, 1500000};
+        static Histogram::MetricFamily & s3_connect = Histogram::Factory::instance().registerMetric(
+            "s3_connect_microseconds",
+            "Time to establish connection with S3, in microseconds.",
+            connect_buckets,
+            {}
+        );
+        s3_connect.withLabels({}).observe(latency);
+
+        if (for_disk_s3)
+        {
+            static Histogram::MetricFamily & disk_s3_connect = Histogram::Factory::instance().registerMetric(
+                "disk_s3_connect_microseconds",
+                "Time to establish connection with DiskS3, in microseconds.",
+                connect_buckets,
+                {}
+            );
+            disk_s3_connect.withLabels({}).observe(latency);
+        }
+        return;
+    }
 
-    static const LatencyBuckets::Event disk_s3_events_map[static_cast<size_t>(S3LatencyType::EnumSize)][static_cast<size_t>(S3MetricKind::EnumSize)] = {
-        {LatencyBuckets::DiskS3FirstByteReadAttempt1Microseconds, LatencyBuckets::DiskS3FirstByteWriteAttempt1Microseconds},
-        {LatencyBuckets::DiskS3FirstByteReadAttempt2Microseconds, LatencyBuckets::DiskS3FirstByteWriteAttempt2Microseconds},
-        {LatencyBuckets::DiskS3FirstByteReadAttemptNMicroseconds, LatencyBuckets::DiskS3FirstByteWriteAttemptNMicroseconds},
-        {LatencyBuckets::DiskS3ConnectMicroseconds, LatencyBuckets::DiskS3ConnectMicroseconds},
-    };
+    const String attempt_label = [](S3LatencyType t)
+    {
+        switch (t)
+        {
+            case S3LatencyType::FirstByteAttempt1: return "1";
+            case S3LatencyType::FirstByteAttempt2: return "2";
+            case S3LatencyType::FirstByteAttemptN: return "N";
+            default: return "UNKNOWN";
+        }
+    }(type);
 
-    S3MetricKind kind = getMetricKind(request);
+    const String http_method_label = [](Aws::Http::HttpMethod m)
+    {
+        switch (m)
```

Comment:
We should return something suitable for `default:` path. Perhaps `UNKNOWN` or just empty string.